internal/db/db.go

```diff
@@ -0,0 +1,68 @@
+package db
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/canonical/chisel/internal/jsonwall"
+	"github.com/klauspost/compress/zstd"
+)
+
+const schema = "0.1"
+
+// New creates a new Chisel DB writer with the proper schema.
+func New() *jsonwall.DBWriter {
+	options := jsonwall.DBWriterOptions{Schema: schema}
+	return jsonwall.NewDBWriter(&options)
+}
+
+func dbPath(root string) string {
+	return filepath.Join(root, ".chisel.db")
+}
+
+// Save uses the provided writer dbw to write the Chisel DB into the standard
+// path under the provided root directory.
+func Save(dbw *jsonwall.DBWriter, root string) error {
+	f, err := os.OpenFile(dbPath(root), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	// chmod the existing file
+	if err := f.Chmod(0644); err != nil {
+		return err
+	}
+	zw, err := zstd.NewWriter(f)
+	if err != nil {
+		return err
+	}
+	if _, err := dbw.WriteTo(zw); err != nil {
+		return err
+	}
+	zw.Close()
+	return nil
+}
+
+// Load reads a Chisel DB from the standard path under the provided root
+// directory. If the Chisel DB doesn't exist, the returned error satisfies
+// os.IsNotExist(err).
+func Load(root string) (*jsonwall.DB, error) {
+	f, err := os.Open(dbPath(root))
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	zr, err := zstd.NewReader(f)
```

Comment:
Per docs:

> To release the resources, you must call the Close() function on a decoder. After this it can no longer be reused, but all running goroutines will be stopped. So you must use this if you will no longer need the Reader."